packages/yew/src/dom_bundle/dom_position.rs

```diff
@@ -0,0 +1,215 @@
+}
+
+impl std::fmt::Debug for DomPosition {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "DomPosition {{ next_sibling: {:?} }}",
+            self.get().map(|n| crate::utils::print_node(&n))
+        )
+    }
+}
+
+impl std::fmt::Debug for RetargetableDomPosition {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:#?}", *self.target.borrow())
+    }
+}
+
+#[cfg(debug_assertions)]
+thread_local! {
+    // A special marker element that should not be referenced
+    static TRAP: Node = gloo::utils::document().create_element("div").unwrap().into();
+}
+
+impl DomPosition {
+    /// Denotes the position just before the given node in its parent's list of children.
+    pub fn at(node: Node) -> Self {
+        Self::create(Some(node))
+    }
+
+    /// Denotes the position at the end of a list of children. The parent is implicit.
+    pub fn at_end() -> Self {
+        Self::create(None)
+    }
+
+    pub fn create(node: Option<Node>) -> Self {
+        Self {
+            variant: DomPositionVariant::Node(node),
+        }
+    }
+
+    /// A new "placeholder" node ref that should not be accessed
+    #[inline]
+    pub fn new_debug_trapped() -> Self {
+        #[cfg(debug_assertions)]
+        {
+            Self::at(TRAP.with(|trap| trap.clone()))
+        }
+        #[cfg(not(debug_assertions))]
+        {
+            Self::at_end()
+        }
+    }
+
+    /// Get the [Node] that comes just after the position, or `None` if this denotes the position at
+    /// the end
+    pub fn get(&self) -> Option<Node> {
+        #[allow(clippy::let_and_return)]
+        let node = match &self.variant {
+            DomPositionVariant::Node(ref n) => n.clone(),
+            DomPositionVariant::Chained(ref chain) => chain.get(),
+        };
+
+        #[cfg(debug_assertions)]
+        TRAP.with(|trap| {
+            assert!(
+                node.as_ref() != Some(trap),
+                "Should not use a trapped node ref. Please report this as an internal bug in yew."
+            )
+        });
+        node
+    }
+}
+
+impl RetargetableDomPosition {
+    pub fn new(initial_position: DomPosition) -> Self {
+        Self {
+            target: Rc::new(RefCell::new(initial_position)),
+        }
+    }
+
+    pub fn new_debug_trapped() -> Self {
+        Self::new(DomPosition::new_debug_trapped())
+    }
+
+    /// Change the [DomPosition] that is targeted. Getting the node from previously obtained
+    /// positions from [`Self::as_position`] will subsequently reflect the result of
+    /// `next_position.get()`.
+    pub fn retarget(&self, next_position: DomPosition) {
+        // TODO: is not defensive against accidental reference loops
+        *self.target.borrow_mut() = next_position;
+    }
+
+    /// Get a [DomPosition] that gets automatically updated when `self` gets retargeted.
+    pub fn as_position(&self) -> DomPosition {
```

Comment:
```suggestion
    pub fn to_position(&self) -> DomPosition {
```

I think it's better to name it `to_position`.
`as_` is usually reserved for a signature that yields a reference.